pkg/controller/machine.go

```diff
@@ -644,39 +669,38 @@ func (c *Controller) updateMachineStatus(machine *machinev1alpha1.Machine, node
 	if err != nil {
 		return fmt.Errorf("failed to get node reference for %s : %v", node.Name, err)
 	}
-	if !equality.Semantic.DeepEqual(machine.Status.NodeRef, ref) {
-		machine.Status.NodeRef = ref
-		updated = true
-	}
-
-	if machine.Status.Versions == nil {
-		machine.Status.Versions = &machinev1alpha1.MachineVersionInfo{}
-	}
 
-	if node.Status.NodeInfo.ContainerRuntimeVersion != "" {
-		runtimeName, runtimeVersion, err = parseContainerRuntime(node.Status.NodeInfo.ContainerRuntimeVersion)
-		if err != nil {
-			glog.V(2).Infof("failed to parse container runtime from node %s: %v", node.Name, err)
-			runtimeName = "unknown"
-			runtimeVersion = "unknown"
+	machine, err = c.updateMachine(machine.Name, func(m *machinev1alpha1.Machine) {
+		if !equality.Semantic.DeepEqual(m.Status.NodeRef, ref) {
```

Comment:
Only put the logic to modify the machine object into the update function.
Otherwise we'll risk triggering an update without doing any changes - which could lead to triggering an update call during each sync.